File: 424-longest-repeating-character-replacement/longest-repeating-character-replacement.go
func characterReplacement(str string, k int) int {
    set := make(map[byte]int)
	left := 0
	right := 0
	currentMaxChar := 0
	result := 0

	for right < len(str) {
		set[str[right]]++

		if set[str[currentMaxChar]] < set[str[right]] {
			currentMaxChar = right
		}

		if (right - left + 1) - set[str[currentMaxChar]] > k {
			set[str[left]]--
			left++
		}

		result = max(result, right - left + 1)
        right++
	}
    return result
}

File: 424-longest-repeating-character-replacement/longest-repeating-character-replacement.java
class Solution {
    public int characterReplacement(String s, int k) {
        int left = 0;
        int right = 0;
        int result = 0;
        int currentMax = 0;
        int[] freq = new int[26];

        while(right < s.length())
        {
            int fr = s.charAt(right) - 'A';
            freq[fr]++;
            currentMax = Math.max(currentMax, freq[fr]);

            while((right - left + 1) - currentMax > k)
            {
                freq[s.charAt(left) - 'A']--;
                left++;
            }
            result = Math.max(result, (right - left + 1));
            right++;
        }
        return result;
    }
}